librz/analysis/var_global.c

```diff
@@ -0,0 +1,419 @@
+	size_t index;
+	rz_vector_lower_bound(&glob->accesses, offset, index, var_access_cmp);
+	RzAnalysisVarAccess *acc = NULL;
+	if (index < glob->accesses.len) {
+		acc = rz_vector_index_ptr(&glob->accesses, index);
+	}
+	if (!acc || acc->offset != offset) {
+		acc = rz_vector_insert(&glob->accesses, index, NULL);
+		acc->offset = offset;
+		acc->type = 0;
+	}
+
+	acc->type |= (ut8)access_type;
+	acc->stackptr = stackptr;
+	acc->reg = rz_str_constpool_get(&analysis->constpool, reg);
+}
+
+/**
+ * \brief Remove the access at the address of global variable
+ * 
+ * \param glob Global variable instance
+ * \param address Where to delete
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_remove_access_at(RzAnalysisVarGlobal *glob, ut64 address) {
+	rz_return_if_fail(glob);
+	st64 offset = (st64)address - (st64)glob->addr;
+	size_t index;
+	rz_vector_lower_bound(&glob->accesses, offset, index, var_access_cmp);
+	if (index >= glob->accesses.len) {
+		return;
+	}
+	RzAnalysisVarAccess *acc = rz_vector_index_ptr(&glob->accesses, index);
+	if (acc->offset == offset) {
+		rz_vector_remove_at(&glob->accesses, index, NULL);
+	}
+}
+
+/**
+ * \brief Clear all the accesses of global variable
+ * 
+ * \param glob Global variable instance
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_clear_accesses(RzAnalysisVarGlobal *glob) {
+	rz_return_if_fail(glob);
+	rz_vector_clear(&glob->accesses);
+}
+
+/**
+ * \brief Add a constaint to global variable
+ * 
+ * \param glob Global variable instance
+ * \param constraint RzTypeConstraint
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_add_constraint(RzAnalysisVarGlobal *glob, RzTypeConstraint *constraint) {
+	rz_return_if_fail(glob && constraint);
+	rz_vector_push(&glob->constraints, constraint);
+}
+
+/**
+ * \brief Get the pritable string of global variable constraints
+ * 
+ * \param glob Global variable instance
+ * \return char *
+ */
+RZ_API char *rz_analysis_var_global_get_constraints_readable(RzAnalysisVarGlobal *glob) {
+	size_t n = glob->constraints.len;
+	if (!n) {
+		return NULL;
+	}
+	bool low = false, high = false;
+	RzStrBuf sb;
+	rz_strbuf_init(&sb);
+	size_t i;
+	for (i = 0; i < n; i += 1) {
+		RzTypeConstraint *constr = rz_vector_index_ptr(&glob->constraints, i);
+		switch (constr->cond) {
+		case RZ_TYPE_COND_LE:
+			if (high) {
+				rz_strbuf_append(&sb, " && ");
+			}
+			rz_strbuf_appendf(&sb, "<= 0x%" PFMT64x "", constr->val);
+			low = true;
+			break;
+		case RZ_TYPE_COND_LT:
+			if (high) {
+				rz_strbuf_append(&sb, " && ");
+			}
+			rz_strbuf_appendf(&sb, "< 0x%" PFMT64x "", constr->val);
+			low = true;
+			break;
+		case RZ_TYPE_COND_GE:
+			rz_strbuf_appendf(&sb, ">= 0x%" PFMT64x "", constr->val);
+			high = true;
+			break;
+		case RZ_TYPE_COND_GT:
+			rz_strbuf_appendf(&sb, "> 0x%" PFMT64x "", constr->val);
+			high = true;
+			break;
+		default:
+			break;
+		}
+		if (low && high && i != n - 1) {
```

Comment:
Most of the jump conditions are comparing the size of the arguments, but there are some special conditions like `JC/JO/JS`, they don't affect the result of constraints because the constraint is the value of the variable.